### PR TITLE
Add binding capabilities to the order flow. Use Angular PF wizard.

### LINKED ISSUE #232: Overlay close button too close to wizard step text on mobile
openshift catalog components 2017-05-17 07-25-12

@jeff-phillips-18 I suspect when we add a fourth step for bind, they won't fit anymore on mobile.

### PR DESCRIPTION
**Pull request by jeff-phillips-18**
Bumps origin-web-common version to v0.0.24

Fixes #232

### CODE REVIEW

File: dist/less/overlay-panel.less [hunk 1]
```diff
@@ -89,6 +103,9 @@ body.overlay-open {
     height: @wizard-pf-steps-indicator-height;
     margin: @wizard-pf-steps-indicator-margin-top-bottom 0;
     padding: 0;
+    @media(max-width: @screen-sm-min) {
+      margin-top: calc(@wizard-pf-steps-indicator-margin-top-bottom + 10px);
```
Comment (rhamilto) [inline dist/less/overlay-panel.less]: Rather than adding more margin to accommodate the close button (which is why I assume you're doing this), how about we move the close closer to the edges of the modal, thus saving valuable vertical screen real estate?

```
.catalogs-overlay-panel-close {
  @media(max-width: @screen-sm-min) {
    right: 15px;
    top: 10px;
  }
}
```

looks like it works.

screen shot 2017-05-17 at 9 58 44 am

File: dist/less/overlay-panel.less [hunk 2]
```diff
@@ -132,3 +149,7 @@ body.overlay-open {
     animation: modalSlideUp 0.2s ease-out forwards;
   }
 }
+
+.wizard-overlay .catalogs-overlay-panel {
+  padding-bottom: 0;
```
Comment (rhamilto) [inline dist/less/overlay-panel.less]: I don't think you need to override here.  I think you just need to remove padding-bottom on .catalog-overlay-panel?

File: dist/less/order-service.less [hunk 3]
```diff
@@ -170,17 +175,24 @@
...
   @media (min-width: @screen-sm-min) {
-    display: flex;
-    min-height: 100%;
-    padding-bottom: (@grid-gutter-width / 2);
-    padding-top: (@grid-gutter-width / 2);
-    .scroll-shadows-horizontal-covers();
-    _::-webkit-full-page-media, _:future, :root & { // only target Safari
-      background: none;
-    }
+    height: calc(@order-service-page-height + @grid-gutter-width);
```
Comment (spadgett) [inline dist/less/order-service.less]: I know we've had some debate on this and I understand why height/max-height are added, but I agree with @rhamilto. As a user, only being able to use a small portion of my screen to fill out what could be a complicated form is frustrating. We don't control what parameters come back from the brokers, and the form will get longer when we add parameters in #231. I feel like the minor inconvenience of the dialog resizing and the buttons moving is the lesser evil, particularly when our wizard only has a few steps.

Comment (spadgett) [inline dist/less/order-service.less]: We can also add a reasonable min-height at non-mobile screen sizes so that it's effectively a fixed height unless there's a lot of content causing it to grow.

Comment (jeff-phillips-18) [inline dist/less/order-service.less]: @serenamarie125 @beanh66 @cshinn Thoughts?

Comment (serenamarie125) [inline dist/less/order-service.less]: From an experience perspective, having the wizard change sizes will be unexpected as well as change the location of the buttons.  Based on my understanding of users and use cases, and applying the 80/20 rule, I would suggest that the behavior of the wizard be that the page size is consistent, and we use scrolling as needed.

Whether that means we need to increase the minimum height to make sure that our "typical" number of parameters is shown without a scroll is a different question.  

@jeff-phillips-18 @rhamilto @spadgett

Comment (spadgett) [inline dist/less/order-service.less]: > Based on my understanding of users and use cases, and applying the 80/20 rule, I would suggest that the behavior of the wizard be that the page size is consistent, and we use scrolling as needed.

Most of our templates have dozens of parameters. If we're applying the 80/20 rule, the common case is to have a lot, at least for the template broker. It's hard to say what other brokers will do.

> Whether that means we need to increase the minimum height to make sure that our "typical" number of parameters is shown without a scroll is a different question.

Even our full screen template page has a scrollbar. I don't think there is a minimum height that avoids it.

As someone who uses this dialog every day: having it resize doesn't bother me, but not being able to use the screen space makes it hard to use (see #177). I can only speak for myself, though.

Comment (serenamarie125) [inline dist/less/order-service.less]: @spadgett dozens of parameters are within the advanced options, or are regular options?  I thought that you and Jessica asked us to get customer feedback on how many of these options are actually specified upfront, implying that we think we are offering too many parameters.  Did i misunderstand that?

From a usability perspective, i still disagree and think that the wizard should remain a consistent size.  @cshinn @beanh66 thoughts?

Comment (beanh66) [inline dist/less/order-service.less]: @serenamarie125 Personally I would rather deal with scrolling than deal with the modal size changing and buttons moving around. One main advantage of a wizard is the consistent layout. I agree we do want to show as much as possible, so changing the minimum height makes sense, but as @spadgett mentioned, users often have to scroll regardless.

**Comment by jwforres** [comment]
On Wed, May 17, 2017 at 2:24 PM, Serena Doyle <notifications@github.com>
wrote:

> *@serenamarie125* commented on this pull request.
> ------------------------------
>
> In dist/less/order-service.less
> <https://github.com/openshift/origin-web-catalog/pull/233#discussion_r117076082>
> :
>
> >    @media (min-width: @screen-sm-min) {
> -    display: flex;
> -    min-height: 100%;
> -    padding-bottom: (@grid-gutter-width / 2);
> -    padding-top: (@grid-gutter-width / 2);
> -    .scroll-shadows-horizontal-covers();
> -    _::-webkit-full-page-media, _:future, :root & { // only target Safari
> -      background: none;
> -    }
> +    height: calc(@order-service-page-height + @grid-gutter-width);
>
> @spadgett <https://github.com/spadgett> dozens of parameters are within
> the advanced options, or are regular options? I thought that you and
> Jessica asked us to get customer feedback on how many of these options are
> actually specified upfront, implying that we think we are offering too many
> parameters. Did i misunderstand that?
>

We were asking about creating a new app from a source builder image, that's
a different question because we have total control over what the form
contains.

Templates are like any other provisioned services in that we have zero
control over what people install into their cluster. Yes most of our OOTB
origin templates can be created with no additional input from the user.
But I can't say if the same is true for all of the JBoss and middleware
templates that ship with OCP, would need to check.  And customers can do
anything.  The parameter schema is defined on the service, we auto-generate
the UI form off of that.

File: dist/less/order-service.less (hunk 3 above)
Comment (spadgett) [inline dist/less/order-service.less]: > dozens of parameters are within the advanced options, or are regular options? I thought that you and Jessica asked us to get customer feedback on how many of these options are actually specified upfront, implying that we think we are offering too many parameters. Did i misunderstand that?

There is no advanced options page for service instance parameters (yet anyway). For context, there are three different forms we show in the overlay from the landing page catalog currently:

1. Builder image forms. We have control over what's shown here. This is the from from #27 that we wanted customer feedback on.
2. Template forms. We **don't** control what the parameters are.
3. Service instance forms (provisioning services from the catalog). We **don't** control what the parameters are.

For (1) and (2), we've reused existing pages in the web console to add the advanced options link. We have not done that for (3), and we don't have an existing page to reuse.

(2) will go away after tech preview when we enable the template broker, so the templates will then show up like other service instances (3) and we'll have to revisit how we deal with parameters.

In short, we've (sort of) solved it for 2 of the 3 cases, but I'm not sure we have a good solution for service instance parameters. If we need to create a standalone provision service page for advanced options, we should open a Trello card because that's another chunk of work that isn't being tracked.

File: src/filters/Applications.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export function applicationHasDeploymentFilter ($filter: any) {
```
Comment (jwforres) [inline src/filters/Applications.ts]: lets not recreate these filters in catalog with slightly different names, I don't want this in multiple repos, its going to be a pain to track down if it ever changes. Move the filters up to common if needed.

File: dist/less/order-service.less (hunk 3 above)
Comment (jeff-phillips-18) [inline dist/less/order-service.less]: I'm really torn on this. Options:

1. Make the wizard resize per page

  This jumps the wizard height around even when selecting options on a single page. This is a bit jarring to me and the next/back buttons move up and down while trying to navigate quickly.

2. Keep a consistent height as it is now

This means more scrolling on pages with lots of parameters. Worse when there are a moderate amount and I have to scroll for one or two parameters when I wouldn't otherwise have to scroll at all. When there are a slew of them, I'm scrolling either way.

3. Keep a consistent height, but use as much window vertical space as possible

Less scrolling for sure, can probably fit an extra 8-10 parameters before needing to scroll. When simply selecting an existing project, the page is largely empty and lots of mouse movement to get down to the next button.

File: dist/less/overlay-panel.less (hunk 1 above)
Comment (jeff-phillips-18) [inline dist/less/overlay-panel.less]: But if there are more steps then eventually it runs over. Will happen in non-mobile as well if there are lots of steps so I was leaning more towards always having the extra margin so that it would never overlap.

File: dist/less/overlay-panel.less (hunk 2 above)
Comment (jeff-phillips-18) [inline dist/less/overlay-panel.less]: I've assumed that would still be wanted by other uses (non-wizard). I'm OK with this if we feel it is OK to change the generic panel and accommodate the change in other uses if necessary.

File: src/components/create-from-builder/create-from-builder.controller.ts [hunk 5]
```diff
@@ -233,4 +224,192 @@ export class CreateFromBuilderController implements angular.IController {
     let name = this.ctrl.imageStream.resource.metadata.name + ":" + this.ctrl.istag.name;
     return this.DataService.get("imagestreamtags", name, { namespace: 'openshift' });
   }
+
+  private preselectService() {
```
Comment (jeff-phillips-18) [inline src/components/create-from-builder/create-from-builder.controller.ts]: Agreed, however the two were different enough that I did not create a binding service. I can add an issue to create a service in common to handle this.

File: dist/less/overlay-panel.less (hunk 1 above)
Comment (rhamilto) [inline dist/less/overlay-panel.less]: IMO, the steps area should be allowed to expand if the steps begin to stack.  No amount of extra margin is going to accommodate and indeterminate number of steps.

File: src/components/order-service/order-service.controller.ts [hunk 6]
```diff
@@ -28,80 +40,114 @@ export class OrderServiceController implements angular.IController {
...
+    this.ctrl.nextTitle = 'Create';
+    this.ctrl.steps[3].allowed = this.ctrl.steps[2].valid;
 
-  public previousStep() {
-    let step = this.ctrl.steps[this.currentStepIndex - 1];
-    this.gotoStep(step);
-  }
+    this.validityWatcher = this.$scope.$watch("$ctrl.forms.bindForm.$valid", (isValid: any, lastValue: any) => {
+      this.ctrl.steps[2].valid = isValid;
+      this.ctrl.steps[3].allowed = this.ctrl.steps[2].valid;
```
Comment (jeff-phillips-18) [inline src/components/order-service/order-service.controller.ts]: Agreed. The indexing got used more and more as I continued to develop and I didn't look back over it. Makes good sense, I will update.

File: src/components/create-from-builder/create-from-builder.controller.ts [hunk 7]
```diff
@@ -233,4 +224,192 @@ export class CreateFromBuilderController implements angular.IController {
...
+          return  rightCreated.localeCompare(leftCreated);
+        } else {
+          return leftReady ? -1 : 1;
+        }
+      });
+      this.ctrl.serviceInstances = instances;
+    }
+  };
+
+  private updateSteps() {
```
Comment (jeff-phillips-18) [inline src/components/create-from-builder/create-from-builder.controller.ts]: We set the title when we enter the page, updateSteps is slightly misnamed as it only occurs when the project selection changes. I'll try to simplify.

File: src/components/order-service/order-service.controller.ts [hunk 8]
```diff
@@ -199,6 +364,9 @@ export class OrderServiceController implements angular.IController {
 
       this.ctrl.orderComplete = readyCondition && readyCondition.status === 'True';
       this.ctrl.error = _.find(conditions, {type: "ProvisionFailed"});
+      if (this.ctrl.shouldBindToApp !== 'none' && this.ctrl.orderComplete && !this.ctrl.error) {
```
Comment (jeff-phillips-18) [inline src/components/order-service/order-service.controller.ts]: OK, great. I didn't know the binding could happen in parallel.

**Comment by jwforres** [comment]
Yep the service catalog backend will wait until the provision completes
before starting the bind operation, the binding object will just stay in a
"not ready" state until then.

**Comment by jeff-phillips-18** [comment]
Marked as WIP since this will require an update to origin-common to include https://github.com/openshift/origin-web-common/pull/73

**Comment by jeff-phillips-18** [comment]
two remaining issues:

wizard height
Firefox stupid long name messes up two column layout.

**Comment by spadgett** [comment]
I won't block the PR on wizard height.

> Firefox stupid long name messes up two column layout.

@rhamilto I know you had this working. Do you know what change might have caused the problem?

**Comment by jeff-phillips-18** [comment]
I figured it out. I had to set wizard-pf-contents width=100%

**Comment by jeff-phillips-18** [comment]
I think adding another 100px to the wizard height is a reasonable compromise for now. Once we see it with lots of parameters it can be easily adjusted if desired.

**Comment by jeff-phillips-18** [comment]
I believe all comments are addressed now. Please take another peek.

Just waiting on the updated version of origin-common now.

**Comment by jeff-phillips-18** [comment]
Removed WIP, bumped origin-web-common to v0.0.24

File: dist/less/order-service.less (hunk 3 above)
Comment (rhamilto) [inline dist/less/order-service.less]: If the decision is to go with a fixed height, there is no need to additionally specify a max-height and a min-height.

Comment (jeff-phillips-18) [inline dist/less/order-service.less]: The min-height prevents the wizard from getting too small if the user resizes the window to be very short. The max-height is the standard height. So, its not *really fixed height.

Comment (rhamilto) [inline dist/less/order-service.less]: No need to use calc on height since this is within less?

Comment (jeff-phillips-18) [inline dist/less/order-service.less]: removed the calc. thanks.

**Comment by jeff-phillips-18** [comment]
@jwforres @rhamilto Any further issues/concerns?

File: src/components/order-service/order-service.controller.ts [hunk 9]
```diff
@@ -164,6 +249,75 @@ export class OrderServiceController implements angular.IController {
     }
   }
 
+  private updateBindability() {
+    this.bindStep.hidden = _.size(this.ctrl.applications) < 1;
```
Comment (jwforres) [inline src/components/order-service/order-service.controller.ts]: Should probably fix this in a follow up PR but there is an existing issue https://github.com/openshift/origin-web-console/issues/1470 around binding even if you have no applications, the bind dialog is incorrect today (it actually shows up always in the overview but gives you the option to pick apps when there are none).  We should always let you create the generic secret binding, even if you don't have applications.

File: src/components/order-service/order-service.controller.ts [hunk 10]
```diff
@@ -183,22 +337,21 @@ export class OrderServiceController implements angular.IController {
...
 
   private watchResults = (resource: any, data: any, context: any) => {
     this.watches.push(this.DataService.watchObject(resource, data.metadata.name, context, (instanceData: any, action: any) => {
       var conditions: any = _.get(instanceData, 'status.conditions');
       var readyCondition: any = _.find(conditions, {type: "Ready"});
 
+      this.ctrl.serviceInstanceName = _.get(instanceData, 'metadata.name');
       this.ctrl.orderComplete = readyCondition && readyCondition.status === 'True';
       this.ctrl.error = _.find(conditions, {type: "ProvisionFailed"});
+      if (this.ctrl.shouldBindToApp !== 'none' && !this.ctrl.bindInProgress) {
```
Comment (jwforres) [inline src/components/order-service/order-service.controller.ts]: I don't think it makes sense to call this from here, its not something you will do more than once, I would just move this to before your call to watchResults.  As long as the creation of the service instance object was successful, which you know at the point where you call watchResults, then you can create the binding.

Comment (jeff-phillips-18) [inline src/components/order-service/order-service.controller.ts]: before the first callback to watch results, we do not know the name of the service instance which is needed to call bindService. I added the check for inProgress to prevent any additional calls to bindService.

File: src/components/order-service/order-service.controller.ts (hunk 9 above)
Comment (jeff-phillips-18) [inline src/components/order-service/order-service.controller.ts]: OK, I can fix this as part of checking if a service is bindable.

File: src/components/order-service/order-service.controller.ts (hunk 10 above)
Comment (jwforres) [inline src/components/order-service/order-service.controller.ts]: You should have the data you need. The success handler to DataService.create gives you the result of the creation (the data arg), which is the updated serviceInstance object that contains the name

Comment (jwforres) [inline src/components/order-service/order-service.controller.ts]: sorry might be the resource arg not the data arg, one of those is the response object

Comment (jeff-phillips-18) [inline src/components/order-service/order-service.controller.ts]: super cool, updated. Thanks!

Comment (jwforres) [inline src/components/order-service/order-service.controller.ts]: I'm not seeing the change, did you push it?
